### PR TITLE
feat(go): add client tests

### PR DESCRIPTION
**Pull request by millotp**
## 🧭 What and Why

🎟 JIRA Ticket: [APIC-709](https://algolia.atlassian.net/browse/APIC-709)

Add clients tests for go, which are not as easy as I thought, and I needed to change a bunch of things in the client too.

[APIC-709]: https://algolia.atlassian.net/browse/APIC-709?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: generators/src/main/java/com/algolia/codegen/cts/tests/TestsClient.java [hunk 1]
```diff
@@ -102,6 +103,10 @@ public void run(Map<String, CodegenModel> models, Map<String, CodegenOperation>
             if (step.expected.error != null) {
               stepOut.put("isError", true);
               stepOut.put("expectedError", step.expected.error);
+              if (language.equals("go") && step.path != null) {
```
Comment (morganleroi) [inline generators/src/main/java/com/algolia/codegen/cts/tests/TestsClient.java:106]: That's why we love Go isn't it ? 🥇

File: clients/algoliasearch-client-go/algolia/utils/utils.go [hunk 2]
```diff
@@ -330,8 +320,8 @@ func (v *NullableTime) UnmarshalJSON(src []byte) error {
 	return json.Unmarshal(src, &v.value)
 }
 
-// isNilorEmpty checks if an input is nil or empty
-func isNilorEmpty(i any) bool {
+// IsNilorEmpty checks if an input is nil or empty.
+func IsNilorEmpty(i any) bool {
```
Comment (morganleroi) [inline clients/algoliasearch-client-go/algolia/utils/utils.go:324]:
```suggestion
func IsNilOrEmpty(i any) bool {
```

?

**Review by morganleroi — APPROVED** [review]
Insane work !!

File: website/docs/contributing/setup-repository.md [hunk 3]
```diff
@@ -39,6 +39,11 @@ The swift images takes a really long time to build (~5 minutes) because of swift
 
 :::
 
+## Tooling
```
Comment (shortcuts) [inline website/docs/contributing/setup-repository.md:42]: tyyy

File: templates/go/configuration.mustache [hunk 4]
```diff
@@ -2,31 +2,12 @@
...
-		client: http.DefaultClient,
-	}
-}
-
-func (r *defaultRequester) Request(req *http.Request) (*http.Response, error) {
-	return r.client.Do(req)
-}
-
 {{#hasRegionalHost}}
+var allowedRegions = [...]string{ {{#allowedRegions}}"{{.}}",{{/allowedRegions}} }
```
Comment (shortcuts) [inline templates/go/configuration.mustache:9]: is there a case where this is an empty slice?

File: templates/go/client.mustache [hunk 5]
```diff
@@ -17,13 +17,14 @@ import (
 	"regexp"
 	"runtime"
 	"strings"
+  "slices"
 	"time"
```
Comment (shortcuts) [inline templates/go/client.mustache:21]:
```suggestion
	"slices"
	"time"
```

File: clients/algoliasearch-client-go/algolia/utils/utils.go (hunk 2 above)
Comment (shortcuts) [reply clients/algoliasearch-client-go/algolia/utils/utils.go:324]: ^

File: clients/algoliasearch-client-go/algolia/transport/configuration.go [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+)
+
+type Configuration struct {
+	AppID  string
+	ApiKey string
+
+	Hosts          []string          `json:"host,omitempty"`
+	DefaultHeader  map[string]string `json:"defaultHeader,omitempty"`
+	UserAgent      string            `json:"userAgent,omitempty"`
+	Debug          bool              `json:"debug,omitempty"`
```
Comment (shortcuts) [inline clients/algoliasearch-client-go/algolia/transport/configuration.go:16]: do we need the json annotations?

File: templates/go/configuration.mustache (hunk 4 above)
Comment (millotp) [reply templates/go/configuration.mustache:9]: after #2561 yes, the logic will have to change

File: templates/go/client.mustache (hunk 5 above)
Comment (millotp) [reply templates/go/client.mustache:21]: it will be formatted anyway, I can't even see the diff in my editor

File: clients/algoliasearch-client-go/algolia/transport/configuration.go (hunk 6 above)
Comment (millotp) [reply clients/algoliasearch-client-go/algolia/transport/configuration.go:16]: maybe not

File: generators/src/main/java/com/algolia/codegen/cts/tests/TestsClient.java (hunk 1 above)
Comment (millotp) [reply generators/src/main/java/com/algolia/codegen/cts/tests/TestsClient.java:104]: so many edge cases !

File: clients/algoliasearch-client-go/algolia/transport/requester.go [hunk 7]
```diff
@@ -22,37 +22,40 @@ const (
...
+			MaxIdleConnsPerHost: DefaultMaxIdleConnsPerHost,
+			Proxy:               http.ProxyFromEnvironment,
+			TLSHandshakeTimeout: DefaultTLSHandshakeTimeout,
+		},
+	}
 }
 
 type Requester interface {
-	Request(req *http.Request) (*http.Response, error)
+	Request(req *http.Request, timout time.Duration, connectTimeout time.Duration) (*http.Response, error)
```
Comment (Fluf22) [inline clients/algoliasearch-client-go/algolia/transport/requester.go:46]:
```suggestion
	Request(req *http.Request, timeout time.Duration, connectTimeout time.Duration) (*http.Response, error)
```

**Review by Fluf22 — COMMENTED** [review]
Sorry, saw a small typo 😬

File: clients/algoliasearch-client-go/algolia/transport/requester.go (hunk 7 above)
Comment (millotp) [reply clients/algoliasearch-client-go/algolia/transport/requester.go:46]: good catch !

**Review by Fluf22 — APPROVED** [review]
Fine by me
